src/v/cloud_storage/tests/delete_records_e2e_test.cc

```diff
@@ -311,3 +311,177 @@ FIXTURE_TEST(test_delete_from_archive_consume, delete_records_e2e_fixture) {
+FIXTURE_TEST(test_delete_from_stm_truncation, delete_records_e2e_fixture) {
+    tests::remote_segment_generator gen(make_kafka_client().get(), *partition);
+    size_t records_per_seg = 5;
+    BOOST_REQUIRE_GE(
+      250,
+      gen.batches_per_segment(records_per_seg)
+        .num_segments(40)
+        .additional_local_segments(10)
+        .produce()
+        .get());
+    archiver->apply_spillover().get();
+    auto& stm_manifest = archiver->manifest();
+    BOOST_REQUIRE_EQUAL(stm_manifest.get_spillover_map().size(), 3);
+    BOOST_REQUIRE_EQUAL(stm_manifest.size(), segs_per_spill);
+    kafka_delete_records_transport deleter(make_kafka_client().get());
+    deleter.start().get();
+
+    // Truncate within the STM at the bound, leaving one segment.
+    auto stm_truncate_offset
+      = archiver->manifest().get_last_kafka_offset().value();
+    deleter
+      .delete_records_from_partition(
+        topic_name,
+        model::partition_id(0),
+        kafka::offset_cast(stm_truncate_offset),
+        5s)
+      .get();
+    // The first housekeeping will cleanup the archive.
+    archiver->housekeeping().get();
+    BOOST_REQUIRE_EQUAL(10, archiver->manifest().size());
+    BOOST_REQUIRE_EQUAL(
+      stm_manifest.get_archive_start_offset(), model::offset{});
+
+    // The next will clear the STM manifest.
+    archiver->housekeeping().get();
+    BOOST_REQUIRE_EQUAL(1, archiver->manifest().size());
+    BOOST_REQUIRE_EQUAL(
+      stm_manifest.get_archive_start_offset(), model::offset{});
+}
+
+// Test that truncation is applied to cloud storage as expected when a
+// DeleteRecords request lands in the archive.
+FIXTURE_TEST(test_delete_from_archive_truncation, delete_records_e2e_fixture) {
+    tests::remote_segment_generator gen(make_kafka_client().get(), *partition);
+    size_t records_per_seg = 5;
+    BOOST_REQUIRE_GE(
+      250,
+      gen.batches_per_segment(records_per_seg)
+        .num_segments(40)
+        .additional_local_segments(10)
+        .produce()
+        .get());
+    archiver->apply_spillover().get();
+    auto& stm_manifest = archiver->manifest();
+    BOOST_REQUIRE_EQUAL(stm_manifest.get_spillover_map().size(), 3);
+    BOOST_REQUIRE_EQUAL(stm_manifest.size(), segs_per_spill);
+    kafka_delete_records_transport deleter(make_kafka_client().get());
+    deleter.start().get();
+
+    // Truncate within the bounds of the first archive segment. Nothing should
+    // be removed.
+    auto& spillover = stm_manifest.get_spillover_map();
+    auto first_seg_commit_offset
+      = *spillover.get_committed_offset_column().at_index(0);
+    auto first_seg_delta_end
+      = *spillover.get_delta_offset_end_column().at_index(0);
+    auto first_seg_last_kafka_offset = first_seg_commit_offset
+                                       - first_seg_delta_end;
+    deleter
+      .delete_records_from_partition(
+        topic_name,
+        model::partition_id(0),
+        model::offset(first_seg_last_kafka_offset),
```

Comment:
It could, I've waffled back and forth between this. The kafka layer has plenty of inconsistency between using kafka:: and model:: for offsets, so I'm going to leave this as is for now.